src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -247,4 +312,78 @@ private boolean changeNodeToken(GraphNode node, TokenMutation tokenMut) {
     }
     return true;
   }
+
+  /**
+   * Given an input graph, roots (as strings), a node map (from string to Graph Nodes), and a
+   * maximum depth, the function returns a graph with only nodes up to a max depth. Edges that don't
+   * contain nodes both reachable up until the max depth are discarded.
+   *
+   * @param graphInput the input graph, as a Mutatable Graph
+   * @param roots the name (string) of the roots
+   * @param graphNodesMap a mapping of strings to GraphNodes
+   * @param maxDepth the maximum depth of a node from a root
+   * @return a graph with nodes only a certain distance from a root
+   */
+  public MutableGraph<GraphNode> getGraphWithMaxDepth(
```

Comment:
The algorithm uses breadth first search from the roots, keeping track of current depth and the elements in this depth. Why not consider depth first search?

Reply:
I actually did not really consider that until you brought it up! I think the BFS just made more sense in my head because I thought of it as layers. 

I added another function right after this one using DFS. I think both have linear complexity and to be completely honest, I'm having a hard time figure out the advantages / disadvantages for both of them relative to each other. 

For both of them right now, we traverse all the edges to find the relevant edges based on the nodes. It doesn't change complexity, but not sure if this could be a bottleneck. Do you think this could be an issue?

Any thoughts would be appreciated!! :)

Reply:
I also don't see a huge advantage/disadvantage for it. I personally find DFS easier to reason about, especially in the case of multiple root nodes. 

Up to you which one you want to keep. You can also ask Ishani and Connor for an opinion!